src/lasp_dependence_dag.erl

```diff
@@ -44,20 +54,53 @@
 %% lets us quickly delete those edges.
 -type process_map() :: dict:dict(pid(), {id(), id()}).
 
+%% Stored metadata for a lasp process.
+%% We don't store the input(s) and output vertices as that information
+%% is implicitly stored by edges in the graph.
+-record(process_metadata, {read :: function(),
+                           transform :: function(),
+                           write :: function()}).
+
+%% We store a mapping Id -> {Pid, Path, MetadataList} to identify the list of
+%% vertices and relationships that get removed as part of a path contraction.
+%%
+%% Used during the cleaving step.
+-type contract_process_id() :: non_neg_integer().
+-type pid_table() :: dict:dict(pid(), contract_process_id()).
+-type optimized_map() :: dict:dict(contract_process_id(),
+                                   {pid(),
+                                    contract_path(),
+                                    list(#process_metadata{})}).
+
 -record(state, {dag :: digraph:graph(),
-                process_map :: process_map()}).
+                process_map :: process_map(),
+                optimized_map :: optimized_map(),
+                pid_table :: pid_table(),
+                contraction_step :: non_neg_integer()}).
 
 %% We store the function metadata as the edge label.
 -record(edge_label, {pid :: pid(),
                     read :: function(),
                     transform :: function(),
                     write :: function()}).
 
+%% @todo Move somewhere else
+%% A tuple of the arguments of a lasp process.
+-type process_args() :: {[{lasp_vertex(), function()}],
+                         function(),
+                         {lasp_vertex(), function()}}.
+
+-record(vertex_label, {process_pointer :: contract_process_id()}).
+
+-type lasp_vertex() :: id() | pid().
```

Comment:
`process_args` is a shorthand for the function metadata in a lasp process, it's used to calculate the hash of a new process, and to restart new ones during the cleaving phase.

`vertex_label` is the label of the removed vertices, and contains a reference to the process that contracted the old path of a vertex.

`lasp_vertex` is a common name for the kinds of vertices in the graph, we can have normal crdt values (`id()`), or process pids for the reads, queries and binds.

I'll add these comments to the file.